File: web/src/core/rng.ts
/* A seeded PRNG.  Determinism is a hard requirement of the program — same sketch plus
 * same edit must give the same result — so every random draw (witness jitter, generic
 * poses for merge decisions, homotopy's gamma trick) comes from a seeded stream here,
 * never from Math.random. */

export class Rng {
  private s: number;

  constructor(seed = 0) {
    this.s = (seed >>> 0) || 0x9e3779b9;
  }

  /** mulberry32 */
  next(): number {
    this.s = (this.s + 0x6d2b79f5) >>> 0;
    let t = this.s;
    t = Math.imul(t ^ (t >>> 15), t | 1);
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61);
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296;
  }

  uniform(lo = 0, hi = 1): number {
    return lo + (hi - lo) * this.next();
  }

  /** Standard normal (Box-Muller; the second variate is kept for the next call). */
  normal(mu = 0, sigma = 1): number {
    if (this.spare !== null) {
      const v = this.spare;
      this.spare = null;
      return mu + sigma * v;
    }
    let u = 0, v = 0, s = 0;
    do {
      u = this.next() * 2 - 1;
      v = this.next() * 2 - 1;
      s = u * u + v * v;
    } while (s === 0 || s >= 1);
    const f = Math.sqrt((-2 * Math.log(s)) / s);
    this.spare = v * f;
    return mu + sigma * u * f;
  }

  private spare: number | null = null;

  int(n: number): number {
    return Math.floor(this.next() * n) % n;
  }

  /** `k` distinct values from 0..n-1, in draw order. */
  sample(n: number, k: number): number[] {
    const pool = Array.from({ length: n }, (_, i) => i);
    const out: number[] = [];
    for (let i = 0; i < k && pool.length; i++) out.push(pool.splice(this.int(pool.length), 1)[0]);
    return out;
  }

  choice<T>(a: T[]): T {
    return a[this.int(a.length)];
  }
}

File: rust/gcs-core/src/rng.rs
//! A seeded PRNG.  Determinism is a hard requirement of the program — same sketch plus same edit
//! must give the same result — so every random draw (witness jitter, generic poses for merge
//! decisions, homotopy's gamma trick) comes from a seeded stream here.

pub struct Rng {
    s: u32,
    spare: Option<f64>,
}

impl Rng {
    pub fn new(seed: u32) -> Rng {
        Rng { s: if seed == 0 { 0x9e37_79b9 } else { seed }, spare: None }
    }

    /// mulberry32.  Named `next` because that is what it is; it is not an `Iterator` — a stream of
    /// f64 with no end has nothing to gain from one.
    #[allow(clippy::should_implement_trait)]
    pub fn next(&mut self) -> f64 {
        self.s = self.s.wrapping_add(0x6d2b_79f5);
        let mut t = self.s;
        t = (t ^ (t >> 15)).wrapping_mul(t | 1);
        t ^= t.wrapping_add((t ^ (t >> 7)).wrapping_mul(t | 61));
        ((t ^ (t >> 14)) as f64) / 4_294_967_296.0
    }

    pub fn uniform(&mut self, lo: f64, hi: f64) -> f64 {
        lo + (hi - lo) * self.next()
    }

    /// Standard normal (Box–Muller; the second variate is kept for the next call).
    pub fn normal(&mut self, mu: f64, sigma: f64) -> f64 {
        if let Some(v) = self.spare.take() {
            return mu + sigma * v;
        }
        loop {
            let u = self.next() * 2.0 - 1.0;
            let v = self.next() * 2.0 - 1.0;
            let s = u * u + v * v;
            if s == 0.0 || s >= 1.0 {
                continue;
            }
            let f = (-2.0 * s.ln() / s).sqrt();
            self.spare = Some(v * f);
            return mu + sigma * u * f;
        }
    }

    pub fn int(&mut self, n: usize) -> usize {
        if n == 0 {
            return 0;
        }
        ((self.next() * n as f64).floor() as usize) % n
    }

    /// `k` distinct values from `0..n`, in draw order.
    pub fn sample(&mut self, n: usize, k: usize) -> Vec<usize> {
        let mut pool: Vec<usize> = (0..n).collect();
        let mut out = Vec::new();
        for _ in 0..k {
            if pool.is_empty() {
                break;
            }
            let i = self.int(pool.len());
            out.push(pool.remove(i));
        }
        out
    }

    pub fn choice<T: Copy>(&mut self, a: &[T]) -> T {
        a[self.int(a.len())]
    }
}
